### PR TITLE
feat(kotlin): add `replaceAllObjects` and `generateSecuredApiKeys` functions

### PR DESCRIPTION
**Pull request by aallam**
## 🧭 What and Why

🎟 JIRA Ticket: DI-1772

### Changes included:

- Add `replaceAllObjects` 
- Add `generateSecuredApiKeys`
- Refactor client options to have access to `Json` for records serialization

### CODE REVIEW

**Review by millotp — COMMENTED** [review]
looks good !

File: clients/algoliasearch-client-kotlin/client/src/commonMain/kotlin/com/algolia/client/extensions/SearchClient.kt [hunk 1]
```diff
@@ -1,32 +1,268 @@
...
+/**
+ * Wait on a delete API ket operation.
+ *
+ * @param maxRetries Maximum number of retry attempts.
+ * @param timeout If specified, the method will throw a
+ *   [kotlinx.coroutines.TimeoutCancellationException] after the timeout value in milliseconds is
+ *   elapsed.
+ * @param requestOptions Additional request configuration.
+ */
+public suspend fun SearchClient.waitKeyDelete(
```
[the remark this answers is not in the document]
Comment (aallam) [reply clients/algoliasearch-client-kotlin/client/src/commonMain/kotlin/com/algolia/client/extensions/SearchClient.kt:128]: Sure, I can add. I also think that passing parameters to a function to have different behavior isn't good API design.

File: clients/algoliasearch-client-kotlin/client/build.gradle.kts [hunk 2]
```diff
@@ -41,5 +41,26 @@ kotlin {
...
+    }
+
+    val jvmTest by getting {
+      dependencies {
+        implementation(libs.kotlin.test.junit)
+        implementation(libs.ktor.client.okhttp)
+      }
+    }
+
+    val appleTest by getting {
```
[the remark this answers is not in the document]
Comment (aallam) [reply clients/algoliasearch-client-kotlin/client/build.gradle.kts:60]: Yup, will fix :D

File: clients/algoliasearch-client-kotlin/client/src/commonMain/kotlin/com/algolia/client/extensions/SearchClient.kt [hunk 3]
```diff
@@ -1,32 +1,268 @@
...
+    records: List<T>,
+    requestOptions: RequestOptions?,
+): Map<String, Long> {
+    if (records.isEmpty()) return emptyMap()
+
+    val requests = records.map { record ->
+        val body = options.json.encodeToJsonElement(serializer, record).jsonObject
+        BatchRequest(action = Action.AddObject, body = body)
+    }
+    val destinationIndex = "${indexName}_tmp_${Random.nextInt(from = 0, until = 100)}"
```
[the remark this answers is not in the document]
Comment (aallam) [reply clients/algoliasearch-client-kotlin/client/src/commonMain/kotlin/com/algolia/client/extensions/SearchClient.kt:204]: I reused the same logic as java/kotlin, does other clients (other then php) have the same pattern ?

File: clients/algoliasearch-client-kotlin/client/src/commonMain/kotlin/com/algolia/client/extensions/internal/Encoding.kt [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+ * Hex utilities.
+ */
+internal object Hex {
+
+    private const val HEX_CODE = "0123456789abcdef"
+
+    /**
+     * Take the [ByteArray] and convert it to a hex encoded [String].
+     */
+    internal fun buildString(byteArray: ByteArray): String {
```
[the remark this answers is not in the document]
Comment (aallam) [reply clients/algoliasearch-client-kotlin/client/src/commonMain/kotlin/com/algolia/client/extensions/internal/Encoding.kt:18]: This is a little bit the ugly part of Kotlin Multiplatform, you either need to:
- Use APIs available for all platforms, or write code that is platform agnostic (this case here), or
- Implement it in your own on each platform (with expect/actual, as done for encoding for example)

File: clients/algoliasearch-client-kotlin/client/src/commonMain/kotlin/com/algolia/client/extensions/SearchClient.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (millotp) [reply clients/algoliasearch-client-kotlin/client/src/commonMain/kotlin/com/algolia/client/extensions/SearchClient.kt:128]: I agree, maybe we should add all the functions to the other languages too, but for compatibility having a single function is nice

File: clients/algoliasearch-client-kotlin/client/src/commonMain/kotlin/com/algolia/client/extensions/SearchClient.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (millotp) [reply clients/algoliasearch-client-kotlin/client/src/commonMain/kotlin/com/algolia/client/extensions/SearchClient.kt:204]: PHP is the only one for now in api clients, I didn't look at the other legacy clients 😢

File: clients/algoliasearch-client-kotlin/client/src/commonMain/kotlin/com/algolia/client/extensions/SearchClient.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (aallam) [reply clients/algoliasearch-client-kotlin/client/src/commonMain/kotlin/com/algolia/client/extensions/SearchClient.kt:194]: Fair enough, I've added the function :)

File: clients/algoliasearch-client-kotlin/client/src/commonMain/kotlin/com/algolia/client/extensions/SearchClient.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (aallam) [reply clients/algoliasearch-client-kotlin/client/src/commonMain/kotlin/com/algolia/client/extensions/SearchClient.kt]: I took a look at legacy clients: [JS](https://github.com/algolia/algoliasearch-client-javascript/blob/252228ab7c5d5c94890885e8ce9bb9b072905756/packages/client-search/src/methods/index/replaceAllObjects.ts#L49), [Java](https://github.com/algolia/algoliasearch-client-java/blob/5671667bbab2a5941594cc3dd653626ce463edb1/algoliasearch-core/src/main/java/com/algolia/search/SearchIndex.java#L1350) and [C#](https://github.com/algolia/algoliasearch-client-csharp/blob/de915f09f6be18a11d7019a2864bd0cd532290fd/src/Algolia.Search/Clients/SearchIndex.cs#L208), and it seems like _almost_ the same pattern is used

**Review by millotp — APPROVED** [review]
thanks !

**Pull request merged**